### ISSUE TITLE
no gl API is called when run opengl tests such as bin/opengl_copy_pixels

### ISSUE THREAD

**Issue by guoyejun** [halide/Halide#2601]
i'm starting from code (912ecef45c0d8f054c04cc5ed7163d1d1013ec5f) on ubuntu 16.04 x64 system,

Here is my command for the build:
HL_TARGET=host-opengl  make test_opengl 
(actually, i tried 'make test_opengl' and the same phenomenon)

And the build finishes with an error, i copy part of the log here. The error might due to the underlying ogl driver, just ignore it now.
g++ -std=c++11 -g -fno-omit-frame-pointer -fno-rtti -I /work/va/Halide/tools -Wall -Werror -Wno-unused-function -Wcast-qual -Wignored-qualifiers -Wno-comment -Wsign-compare -Wno-unknown-warning-option -Wno-psabi -O0 /work/va/Halide/test/opengl/rewrap_texture.cpp -Iinclude -I/work/va/Halide/src -Lbin -lHalide -ldl -lpthread -lz -rdynamic -Wl,--rpath=/work/va/Halide/bin -lGL -o bin/opengl_rewrap_texture -MMD -MF bin/build/test_opengl_rewrap_texture.d
cd bin/build/tmp ; /work/va/Halide/bin/opengl_rewrap_texture 2>&1
Success!

g++ -std=c++11 -g -fno-omit-frame-pointer -fno-rtti -I /work/va/Halide/tools -Wall -Werror -Wno-unused-function -Wcast-qual -Wignored-qualifiers -Wno-comment -Wsign-compare -Wno-unknown-warning-option -Wno-psabi -O0 /work/va/Halide/test/opengl/save_state.cpp -Iinclude -I/work/va/Halide/src -Lbin -lHalide -ldl -lpthread -lz -rdynamic -Wl,--rpath=/work/va/Halide/bin -lGL -o bin/opengl_save_state -MMD -MF bin/build/test_opengl_save_state.d
cd bin/build/tmp ; /work/va/Halide/bin/opengl_save_state 2>&1
realize did not restore element array buffer binding: initial value was 6 (0x6), current value is 0 (0)
realize did not restore boolean VertexAttributeArrayEnabled(0): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(1): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(2): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(3): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(4): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(5): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(6): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(7): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(8): initial value was true, current value is false
realize did not restore boolean VertexAttributeArrayEnabled(9): initial value was true, current value is false
copy_to_host did not restore element array buffer binding: initial value was 10 (0xa), current value is 0 (0)
copy_to_host did not restore boolean VertexAttributeArrayEnabled(0): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(1): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(2): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(3): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(4): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(5): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(6): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(7): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(8): initial value was true, current value is false
copy_to_host did not restore boolean VertexAttributeArrayEnabled(9): initial value was true, current value is false
realize did not restore element array buffer binding: initial value was 12 (0xc), current value is 0 (0)
copy_to_host did not restore element array buffer binding: initial value was 16 (0x10), current value is 0 (0)
Makefile:1414: recipe for target 'opengl_save_state' failed
make: *** [opengl_save_state] Error 1

Anyway, i still get some test successfully built, for example, bin/opengl_lut, bin/opengl_float_texture , bin/opengl_copy_pixels and their test results are all passed.

Then, i step into the test with gdb, i can see that some OpenGL relative libraries are loaded, see below. 
/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
/usr/lib/x86_64-linux-gnu/dri/i965_dri.so

but, it is unable to set breakpoint with glUseProgram, and I also tried with:
(gdb) b opengl.cpp:876
No source file named opengl.cpp.

It makes me guess that opengl.cpp is not built into the project, and Target::OpenGL is always working on host CPU.

not sure if i mis-understood something, thanks.

**Comment by shoaibkamil** [comment]
I believe those tests automatically set the target to include OpenGL.  Because they operate in JIT mode, btw, if you want to modify the target you need to set `HL_JIT_TARGET`.

You can see more explicit output using a debug target, like:
```
$ HL_JIT_TARGET=host-opengl-debug make test_opengl`.
```

From the output above, it does seem like OpenGL calls are occurring, but adding `debug` to the target will give more detailed output (and perhaps help diagnose the error you're seeing)

**Comment by guoyejun** [comment]
thanks shoaibkamil.

i tried HL_JIT_TARGET=host-opengl-debug and I did see the functions such as halide_opengl_copy_to_device are called (i changed the code to add my special log to double check), but it is still unable to set breakpoint for glUseProgram, Halide::Runtime::Internal::OpenGL::halide_opengl_copy_to_device, opengl.cpp:1065, etc. Anyway, it shows that GL APIs should be called.

btw, why these opengl tests (using jit) still PASSED when i set the wrong flag (HL_TARGET=host-opengl), my understanding is that, opengl backend (for jit) is only enabled when set the HL_JIT_TARGET correctly.

**Comment by shoaibkamil** [comment]
> btw, why these opengl tests (using jit) still PASSED when i set the wrong flag (HL_TARGET=host-opengl), my understanding is that, opengl backend (for jit) is only enabled when set the HL_JIT_TARGET correctly.

That's generally correct; the OpenGL backend is only enabled when it is included in the target.  However, for the tests, we explicitly set the target to include OpenGL from within the program.  You can see this e.g. [here](https://github.com/halide/Halide/blob/6a035395962340c3abc28c3c0d99ef3a31673cb7/test/opengl/conv_select.cpp#L12) in one of the tests.

**Comment by zvookin** [comment]
GPU specific tests either set the GPU in the target explicitly, in which
case they can only be run on platforms that have the appropriate GPU
support, or they print a message saying they are skipping running and pass
if the GPU is not specified in the target.

In order to get things to happen on a GPU, one needs the correct GPU API
specified in the target and some scheduling to place code on that GPU in
the schedule. It is not clear to me what the issue here is however.

-Z-

**Comment by guoyejun** [comment]
Here are my steps:
$ rm -rf *
$ git checkout .
$ export LLVM_CONFIG=/work/llvm_dist/llvm50/bin/llvm-config
$ export CLANG=/work/llvm_dist/llvm50/bin/clang
$ export | grep HL
//nothing output
$ make test_opengl
$ ./bin/opengl_lut 
PASSED
$ ./bin/opengl_copy_pixels 
Success!
$ gdb bin/opengl_lut 
(gdb) r
(gdb) info sharedlibrary 
...
0x00007fffef89ae00  0x00007fffef8d95f2  Yes (*)     /usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
0x00007fffec9a83b0  0x00007fffeceaf439  Yes (*)     /usr/lib/x86_64-linux-gnu/dri/i965_dri.so
...

my concern is that: 
Since I built halide without setting HL_JIT_TARGET nor HL_TARGET, and so the OpenGL backend  is not enabled. While the OpenGL tests all set target with Target::OpenGL in there source code (const Target target = get_jit_target_from_environment().with_feature(Target::OpenGL)), my understanding is that these tests should be skipped or failed, but according to my experiment, why they are PASSED? (i did not find that they are skipped both from source code and the log, and the OpenGL libraries are even loaded!)

**Comment by zvookin** [comment]
Re: "Since I built halide without setting HL_JIT_TARGET nor HL_TARGET, and
so the OpenGL backend is not enabled." Those two environment variables
control compilation of Halide code, either for AOT or JIT, but they are not
used in making the compiler (libHalide) itself. If you want to disable
inclusion of the OpenGL backend when compiling Halide, you'll need to put
"WITH_OPENGL=" on the make command line, or something similar for cmake.

The OpenGL specific tests turn on OpenGL in the target. One has to request
the OpenGL specific tests explicitly so they don't run on platforms that do
not have OpenGL support. This design is a bit messy. When we started on
OpenGL, it wasn't clear how much of the test suite it would be able to run
int he shader model and the initial tests covering this were OpenGL
specific.

-Z-
